nixos/ollama: replace flawed sandboxing option

mweinelt opened the pull request:
The ollama module in its default cofniguration relies on systemd's `DynamicUser=` feature for user allocation. In #305076 that allocation was made conditional and tied to the `sandboxing` option, that was intended to fix access to model directories outside the allocated state directory.

However, by disabling sandboxing ollama would inadvertently run as root, given that `User=` and `Group=` are not required to be set.

The correct way to grant access to other paths is to allocate static user and group, and grant permissions to the destination path to that allocation.

We therefore replace the sandboxing option user and group options, that default to `null`, which means they default to `DynamicUser=`, but can be replaced with a statically allocated user/group, with stable uid/gid.

Fixes: 552eb759 ("nixos/ollama: add options to bypass sandboxing")

## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

mweinelt committed:
nixos/ollama: replace flawed sandboxing option

The ollama module in its default configuration relies on systemd's
`DynamicUser=` feature for user allocation. In #305076 that allocation
was made conditional and tied to the `sandboxing` option, that was
intended to fix access to model directories outside the allocated state
directory.

However, by disabling sandboxing ollama would inadvertently run as root,
given that `User=` and `Group=` are not required to be set.

The correct way to grant access to other paths is to allocate static
user and group, and grant permissions to the destination path to that
allocation.

We therefore replace the sandboxing option user and group options, that
default to `null`, which means they default to `DynamicUser=`, but can
be replaced with a statically allocated user/group, and thereby a stable
uid/gid.

Fixes: 552eb759 ("nixos/ollama: add options to bypass sandboxing")

drupol commented on nixos/modules/services/misc/ollama.nix:
```diff
@@ -11,22 +11,47 @@ in
       home = lib.mkOption {
         type = types.str;
         default = "%S/ollama";
+        defaultText = "/var/lib/ollama";
```
TIL, thanks !

drupol reviewed (approved):
Diff LGTM !

abysssol commented on nixos/modules/services/misc/ollama.nix:
```diff
@@ -11,22 +11,47 @@ in
       enable = lib.mkEnableOption "ollama server for local large language models";
       package = lib.mkPackageOption pkgs "ollama" { };
+
+      user = lib.mkOption {
+        type = with types; nullOr str;
+        default = null;
+        example = "ollama";
+        description = ''
+          User account under which to run ollama. Defaults to [`DynamicUser`](https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#DynamicUser=)
+          when set to `null`.
+        '';
+      };
+
+      group = lib.mkOption {
+        type = with types; nullOr str;
+        default = cfg.user;
+        defaultText = literalExpression "services.ollama.user";
+        example = "ollama";
+        description = ''
+          Group under which to run ollama. Only used when `services.ollama.user` is set.
+        '';
+      };
```
Is it necessarily a good idea to make the users specify the user and group? Shouldn't they be defined separately, maybe still using the `sandbox` option to specify when to use them?

Something like
``` nix
users = lib.mkIf (cfg.sandbox == false) {
  users.ollama.isSystemUser = true;
  users.ollama.group = "ollama";
  groups.ollama = {};
};
systemd.services.ollama = {
  # omitted
  serviceConfig = (if cfg.sandbox then {
    DynamicUser = true;
  } else {
    User = "ollama";
    Group = "ollama";
  });
  # omitted
};
```

mweinelt replied:
https://search.nixos.org/options?channel=24.05&from=0&size=50&sort=relevance&type=packages&query=.sandbox

That's frankly not a good nor common name for that kind of thing.

abysssol replied:
The name can be changed (maybe just to `dynamicUser`), but I feel like having to manually set a name for the service's user is a worse user experience than just setting a boolean option, and unnecessarily complex for the end goal (enabling/disabling dynamic user and filesystem sandboxing).

Maybe I'm just missing something though; if you can, please explain the benefits of manually setting the user and group names.

mweinelt replied:
The user/group options are just very common. I would like to stick with the expectations people have picked up frorm other modules.

abysssol replied:
I've never seen it before. But even if it is a well established convention, conventions should only be kept when there is a benefit to doing so, which I don't see here. Is there some circumstance where changing the name of the service's user or group is useful or necessary? If not, then what benefit could there be in requiring the user to specify the name, instead of simply toggling the features that are necessary for their workflow?

mweinelt replied:
I think in most cases where this is not desirable, DynamicUser is just left off.

emilazy replied:
> Is there some circumstance where changing the name of the service's user or group is useful or necessary?

It’s useful if you need the service to run as an existing user that also has access to other things the service will integrate with.

I don’t know if that’s relevant in this case, but e.g. perhaps you want to have your web app in the same group as the service so that it can access generated files. This comes up often with ACME certificate generation, for instance.

I agree that this pattern is very common and reasonable to use when possible. If there’s no reason to ever twiddle these settings, then `DynamicUser` should simply be unconditional instead.

mweinelt replied:
The reason to change the models dir is that models might be larger than your mountpoint below `/var/lib/ollama`.

abysssol replied:
@emilazy See issue  #304531 and  pr #305076 for motivation and context for why I made it possible to disable `DynamicUser` and exposed `ReadWritePaths` as `writablePaths`.

mweinelt committed:
nixos/ollama: remove writablePaths option

Making the models directory always writable is much simpler, than having
to watch out for an option to facilitate that.

mweinelt commented on nixos/modules/services/misc/ollama.nix:
```diff
@@ -159,12 +165,16 @@ in
         StateDirectory = [ "ollama" ];
-        DynamicUser = cfg.sandbox;
-        ReadWritePaths = cfg.writablePaths;
+        ReadWritePaths = [ cfg.models ];
```
Is there any other path that commonly needs to be writable?

abysssol replied:
Not as far as I know, but I'm not the one who needs these (I don't even use ollama anymore). I specifically think that you should leave `writablePaths` as is, and simply have `cfg.models` as default.
``` nix
writablePaths = lib.mkOption {
  type = types.listOf types.str;
  default = [ cfg.models ];
  # omitted
};
```

mweinelt replied:
Again, the ollama module is an outlier here. https://search.nixos.org/options?channel=24.05&size=50&sort=relevance&type=packages&query=writablePaths

There is always the direct way via `systemd.services.ollama.serviceConfig.ReadWritePaths` that works for every service out there.

abysssol replied:
> Again, the ollama module is an outlier here.

That's fine, there's nothing wrong with being an outlier. Remember, before anything becomes a convention, it begins as just a few outliers, which eventually becomes common practice.

If you have genuine reasons why the options should be changed and removed, I'm listening. But I haven't heard much other than "It's not common practice", which isn't a very logical argument.

> There is always the direct way via `systemd.services.ollama.serviceConfig.ReadWritePaths` that works for every service out there.

Yes, but not every nix user knows about that. In fact, I used nix for over a year, maybe two, before I learned that services could be overridden like that. I don't think it's especially common knowledge among non-contributors, which are the vast majority of total nix users.

When I can, I prefer to make an interface that is as simple and convenient as possible while still allowing as many possibilities as are useful.

mweinelt replied:
If the only directory that needs to be writable is the model directory, then it is not a valuable option to have and should be removed.

And looking at part 2 of [RFC42](https://github.com/NixOS/rfcs/blob/45b76f20add8d7e0e1d0bc0eed9357b067c899a6/rfcs/0042-config-option.md#valuable-options), the option is neither of popular, mandatory, or sensitive.

Just because we can expose anything as an option, doesn't mean we should: https://github.com/NixOS/rfcs/blob/45b76f20add8d7e0e1d0bc0eed9357b067c899a6/rfcs/0042-config-option.md#part-2

abysssol replied:
> If the only directory that needs to be writable is the model directory, then it is not a valuable option to have and should be removed.

Fair if true, though that is an assumption that you're making.

@suhr could you give your opinion on this? Is the `model` directory the only place that's likely to be useful to write to through the `dynamicUser` sandbox? Were there any other directories where this was necessary?

mweinelt replied:
> though that is an assumption that you're making.

https://github.com/ollama/ollama/blob/5534f2cc6a3f29022998950472741d16e7a66b40/envconfig/config.go#L85-L100

- `OLLAMA_TMPDIR` should go into `/tmp`, together with `PrivateTmp=true`
- `OLLAMA_RUNNERS_DIR` not sure what that is

abysssol replied:
> `OLLAMA_RUNNERS_DIR` not sure what that is

Basically, ollama compiles various runner binaries (cpu, cpu with simd, cuda, rocm) that are what actually run the ai models, then compresses and embeds these binaries into the main ollama binary at compile time. At runtime, they are decompressed and the gpu is tried first (cuda/rocm), falling back to a runner with fewer requirements (simd/cpu) upon failure.

I'm not sure what exactly `OLLAMA_RUNNERS_DIR` does, but one possibility is that this directory is where ollama extracts these runner binaries before executing them.

See [this comment](https://github.com/NixOS/nixpkgs/pull/323249#discussion_r1660362899) for a more detailed explanation.

mweinelt replied:
I would expect that they are then dumped into the wording directory, because even with hardening it still works for me. They are also probably not significant in size, when compared to the models.

abysssol replied:
> Is there any other path that commonly needs to be writable?

It seems to me that `cfg.home` should also be added to `ReadWritePaths`.

mweinelt replied:
> I would expect that they are then dumped into the working directory

Actually they seem to land in /tmp, which is fine as well.

```
;evel=INFO source=payload.go:30 msg="extracting embedded files" dir=/tmp/ollama375027117/runners
```

> It seems to me that cfg.home should also be added to ReadWritePaths.

Possibly :thinking:

abysssol replied:
If `cfg.home` isn't made writable, any custom home directory (eg `/home/user`) will cause errors and possibly crashes whenever ollama tries to write to any directory that isn't `${cfg.home}/models`.

That seems like a big problem to me, since the user can't make it writable with `writablePaths` anymore, and `systemd.services.ollama.serviceConfig.ReadWritePaths` has very low discoverability.

mweinelt replied:
Thanks for expanding on that, I was focusing on getting amdgpu to work in the meantime.

Added as prposed.

mweinelt committed:
nixos/ollama: harden systemd unit

Tested with CPU and CUDA acceleration. Researched for ROCm, but I have no
compatible card to test it with.

mweinelt commented:
I added hardening for the systemd unit on top and tested it with CPU and CUDA acceleration. Wondering how to set up ROCM, I have an RX5700 card accessible.

mweinelt commented:
```
source=payload.go:44 msg="Dynamic LLM libraries [cpu cpu_avx cpu_avx2 rocm]"
source=gpu.go:205 msg="looking for compatible GPUs"
source=amd_linux.go:58 msg="ollama recommends running the https://www.amd.com/en/support/linux-drivers" error="amdgpu version file missing: /sys/module/amdgpu/version stat /sys/module/amdgpu/version: no such file or directory"
source=amd_linux.go:325 msg="amdgpu is not supported" gpu=0 gpu_type=gfx1010 library=/nix/store/v13vgzd2r1jnhk6y11s15w5n21qkx43d-rocm-path/lib supported_types=[]
source=amd_linux.go:327 msg="See https://github.com/ollama/ollama/blob/main/docs/gpu.md#overrides for HSA_OVERRIDE_GFX_VERSION usage"
source=amd_linux.go:345 msg="no compatible amdgpu devices detected"
source=gpu.go:346 msg="no compatible GPUs were discovered"
```

Okay, it wants the AMDGPU Pro drivers, which were removed from nixpkgs iirc.

abysssol commented:
> Okay, it wants the AMDGPU Pro drivers

I'm not sure about that. I've never needed them before. Have you tested the ollama service without your changes to ensure that you have a working setup? It's possible that your gpu isn't supported by rocm, and you'll need to set `rocmOverrideGfx` to allow it to work.

mweinelt commented:
While 10.3.0 just coredumps ollama, 9.4.0 manages to crash my GPU driver which required a hard reset:

```
amdgpu 0000:03:00.0: amdgpu: HIQ MQD's queue_doorbell_id0 is not 0, Queue preemption time out
amdgpu: Failed to evict process queues
amdgpu: Failed to quiesce KFD
amdgpu 0000:03:00.0: amdgpu: GPU reset begin!
amdgpu: Failed to restore process queues
amdgpu: amdgpu_amdkfd_restore_userptr_worker: Failed to resume KFD
amdgpu: Failed to quiesce KFD
```

Doesn't look like ollama wants to support that generation (RX5000 series) either: https://github.com/ollama/ollama/blob/main/docs/gpu.md#amd-radeon

I assume that we'll at least need to whitelist `/dev/kfd` and maybe `/dev/dri/renderD128`

SomeoneSerge commented on nixos/modules/services/misc/ollama.nix:
`strace ... |& grep /dev` might be helpful

mweinelt replied:
Okay, that leaves me with

- /dev/null
- /dev/random
- /dev/kfd
- /dev/dri/renderD128
- /dev/dri/card128 (ENOENT, why not card1?)

mweinelt replied:
Use wildcards for render and card device nodes.

abysssol commented on nixos/modules/services/misc/ollama.nix:
```diff
@@ -1,71 +1,72 @@
       "Use `services.ollama.host` and `services.ollama.port` instead.")
+    (lib.mkRemovedOptionModule [ "services" "ollama" "sandbox" ]
+      "Set `services.ollama.user` and `services.ollama.group` instead.")
+    (lib.mkRemovedOptionModule [ "services" "ollama" "writablePaths" ]
+      "The models directory is now always writable." )
```
```suggestion
    (lib.mkRemovedOptionModule [ "services" "ollama" "writablePaths" ]
      "The `models` directory is now always writable. To make other directories writable, use `systemd.services.ollama.serviceConfig.ReadWritePaths`." )
```
I think this should mention `systemd.services.ollama.serviceConfig.ReadWritePaths` for anyone that really does need to whitelist any directories other than `models`.

abysssol commented on nixos/modules/services/misc/ollama.nix:
```diff
@@ -1,71 +1,72 @@
       home = lib.mkOption {
         type = types.str;
-        default = "%S/ollama";
+        default = "/var/lib//ollama";
```
Looks like you left a double slash here, I presume accidentally.

Also, why not use `default` and `defaultText` like you did elsewhere?

mweinelt replied:
Because I reuse `home` for users.users.<name>.home and the format string is specific to the systemd unit.

drupol replied:
I don't understand either.

mweinelt replied:
Don't understand what?

drupol replied:
Your explanation on this.

mweinelt replied:
Is it clearer with the excerpt?

drupol replied:
OK, thank you.

abysssol replied:
Then shouldn't `models` be changed to be based on `home` too?

``` diff
 models = lib.mkOption {
   type = types.str;
-  default = "%S/ollama/models";
-  defaultText = "/var/lib/ollama/models";
+  default = "${cfg.home}/models";
+  defaultText = "$HOME/models";
```

mweinelt replied:
Sure, why not.

mweinelt committed:
nixos/doc/rl-2411: add ollama changes

mweinelt committed:
nixos/ollama: make host example dualstack wildcard

Binding to IPv6 wildcard generally binds dual-stack, which is also a
better example because the way to specify it is not always very clear.

mweinelt commented:
@ofborg build ollama.tests

abysssol commented on nixos/modules/services/misc/ollama.nix:
```diff
@@ -159,12 +169,64 @@ in
         HSA_OVERRIDE_GFX_VERSION = lib.mkIf (cfg.rocmOverrideGfx != null) cfg.rocmOverrideGfx;
       };
-      serviceConfig = {
+      serviceConfig = lib.optionalAttrs staticUser {
+        User = cfg.user;
+        Group = cfg.group;
+      } // {
+        DynamicUser = true;
```
When `staticUser == true`, then `DynamicUser` will be enabled as well as `User` and `Group` being set. Are you sure that'll work right?

mweinelt replied:
Yes, that works:

> If a statically allocated user or group of the configured name already exists, it is used and no dynamic user/group is allocated.

https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#DynamicUser=

Atemu reviewed (commented):
This breaks AMDGPU via ROCm unfortunately.

It was arguably broken upstream before but you could make it work using i.e.

```Nix
  services.ollama = {
    enable = true;
    acceleration = "rocm";
    rocmOverrideGfx = "10.3.0"; # depends on the GPU
    environmentVariables = {
      HIP_VISIBLE_DEVICES = "0"; # depends on host setup
    };
  };
```

Before: 

```
level=INFO source=payload.go:44 msg="Dynamic LLM libraries [cpu cpu_avx cpu_avx2 rocm]"
level=INFO source=gpu.go:205 msg="looking for compatible GPUs"
level=WARN source=amd_linux.go:58 msg="ollama recommends running the https://www.amd.com/en/support/linux-drivers" error="amdgpu version file missing: /sys/module/amdgpu/version stat /sys/module/amdgpu/version: no such file or directory"
level=INFO source=amd_linux.go:333 msg="skipping rocm gfx compatibility check" HSA_OVERRIDE_GFX_VERSION=10.3.0
level=INFO source=amd_linux.go:259 msg="unsupported Radeon iGPU detected skipping" id=1 total="64.0 MiB"
level=INFO source=types.go:105 msg="inference compute" id=0 library=rocm compute=gfx1030 driver=0.0 name=1002:73bf total="16.0 GiB" available="15.5 GiB""
```

After:

```
level=INFO source=payload.go:44 msg="Dynamic LLM libraries [cpu cpu_avx cpu_avx2 rocm]"
level=INFO source=gpu.go:205 msg="looking for compatible GPUs"
level=WARN source=gpu.go:212 msg="error looking up system memory" error="open /proc/meminfo: no such file or directory"
level=WARN source=amd_linux.go:58 msg="ollama recommends running the https://www.amd.com/en/support/linux-drivers" error="amdgpu version file missing: /sys/module/amdgpu/version stat /sys/module/amdgpu/version: no such file or directory"
level=WARN source=amd_linux.go:186 msg="amdgpu too old gfx000" gpu=0
level=WARN source=amd_linux.go:186 msg="amdgpu too old gfx000" gpu=1
level=INFO source=amd_linux.go:345 msg="no compatible amdgpu devices detected"
level=INFO source=gpu.go:346 msg="no compatible GPUs were discovered"
level=INFO source=types.go:105 msg="inference compute" id=0 library=cpu compute="" driver=0.0 name="" total="0 B" available="0 B"
```

mweinelt commented:
```
level=WARN source=gpu.go:212 msg="error looking up system memory" error="open /proc/meminfo: no such file or directory"
```

Fixed by setting `ProcSubset = "all"`.

```
level=WARN source=amd_linux.go:186 msg="amdgpu too old gfx000" gpu=0
level=WARN source=amd_linux.go:186 msg="amdgpu too old gfx000" gpu=1
```

Fixed by moving `DeviceAllow=` to device types, rather than paths.

abysssol commented on nixos/modules/services/misc/ollama.nix:
```diff
@@ -1,75 +1,76 @@
         type = types.str;
-        default = "%S/ollama/models";
+        default = "${cfg.home}/models";
+        defaultText = literalExpression "\${config.ervices.ollama.home}/models";
```
Typo: `services` is missing an `s`.

Also, I'm don't think that `literalExpression` is a good fit here, since the value is indeed an interpolated string, not a bare path. I'm pretty sure the `${}` interpolation syntax doesn't even work outside of strings.

abysssol replied:
If you prefer, you could also change it to something like:
``` nix
literalExpression ''config.services.ollama.home + "/models"''
```

mweinelt replied:
There is nothing wrong with interpolation, given that the option uses `type = str`, so unless we are going for `type = path` the change seems not that important.

emilazy replied:
`ervices` is a typo and you’re missing the internal quotes (`''"…"''`).

abysssol replied:
Indeed, I prefer showing it as interpolation, I was simply expanding on potential options.

abysssol replied:
@emilazy  As far as I understand, `literalExpression` simply renders strings as-is, but without the quotes that it adds by default. So using `literalExpression` with internal quotes is simply doing the same thing as not using `literalExpression` at all, but with extra steps (added complexity and cognitive overhead).

emilazy replied:
Oh, true. Never mind me then. I was mixing it up with `literalMD` or whatever it's called now.

Atemu commented on nixos/modules/services/misc/ollama.nix:
```diff
@@ -159,12 +169,63 @@ in
+        ];
+
+        CapabilityBoundingSet = [ "" ];
+        DeviceAllow = [
+          # CUDA
+          # https://docs.nvidia.com/dgx/pdf/dgx-os-5-user-guide.pdf
+          "char-nvidiactl"
+          "char-nvidia-caps"
+          "char-nvidia-uvm"
+          # ROCm
+          "char-drm"
+          "char-kfd"
+        ];
```
This reads a lot nicer too, will test it out on my GPU.

Atemu reviewed (approved):
Works on my machine (6800XT)

abysssol commented on nixos/modules/services/misc/ollama.nix:
```diff
@@ -1,75 +1,76 @@
+      group = lib.mkOption {
+        type = with types; nullOr str;
+        default = cfg.user;
+        defaultText = literalExpression "services.ollama.user";
```
Since `models.defaultText` is set to `"\${config.services.ollama.home}/models";`, shouldn't this be `"config.services.ollama.user"` instead? Otherwise neither of them should have the `config.` prefix.

mweinelt replied:
Yes, fixed.

abysssol reviewed (approved):
Looks pretty good, I didn't find any major issues. I don't understand all of the options set in `serviceConfig`, though; hopefully I didn't miss anything important.

drupol commented:
Great work you all !
